core/src/main/java/org/apache/iceberg/Partitioning.java

```diff
@@ -246,4 +246,45 @@ private static boolean equivalentIgnoringNames(PartitionField field, PartitionFi
   private static boolean compatibleTransforms(Transform<?, ?> t1, Transform<?, ?> t2) {
     return t1.equals(t2) || t1.equals(Transforms.alwaysNull()) || t2.equals(Transforms.alwaysNull());
   }
+
+  /**
+   * Adapts the provided partition data to match the table partition type built using {@link #partitionType(Table)}.
+   *
+   * @param partitionType a table partition type that includes partition fields from all specs
+   * @param spec a partition spec
+   * @param partition a struct containing partition data that belongs to the spec
+   * @return the reconciled partition type
+   */
+  public static StructLike reconcilePartition(StructType partitionType, PartitionSpec spec, StructLike partition) {
```

Comment:
Naming ideas?

Reply:
Should we use struct projections rather than copying values?

Reply:
coercePartition?

Reply:
I'll check our current projections to see if they support going from a more specific to a more general schema.

Reply:
I'll switch to `coercePartition` too.

Reply:
Fixed.